### DISCUSSION TITLE
Frozen models idea/implementation / lazy ghost object

### DISCUSSION THREAD

**Discussion by macropay-solutions** [laravel/framework#53875] [Ideas]
Thanks to @cosmastech for the [idea](https://github.com/laravel/framework/pull/53472#issue-2650658411)

> A frozen model communicates intent: you can look, but you don't touch.

  and to @bulletproof-coding  for this [example.](https://github.com/laravel/framework/pull/53472#issuecomment-2532941622)

We would like to receive some feedback on [this draft + demo.](https://github.com/laravel/framework/pull/53472#issuecomment-2536579988)

**Comment by macropay-solutions** [comment]
We updated again the proposed [solution](https://github.com/macropay-solutions/laravel-crud-wizard-free/pull/8/files) to **Make the stdClass read only also for its content**.
OBS.
The reflection or closure binding will allow changes on the private stdClass but that will not propagate to the model.

**Comment by Sophist-UK** [comment]
The idea of immutable objects has a very sound basis in Computer Science for improving reliability, and indeed Data Transfer Objects are generally immutable and often the more generalised Data Objects are immutable also.

The reasons for this are AFAICS twofold:

1. Objects visible to one part of the system cannot then be later modified by another part of the system creating confusion.
2. Objects cannot be mistakenly updated when they are supposed to be fixed.

In PHP I would guess that an immutable object is essentially a read-only class where all attributes are read-only and set only when the object is constructed.

Creating a new database record will mean creating a new read-only Model passing all attributes when it is constructed. When it is saved and you know the uniqueid, you will then need to create a new immutable object containing all the same attributes plus the id.

I would imagine that sometimes (perhaps even most times) you might want some fields in a model to be "frozen" and others to be updateable.

When you want to update a record, you will need to either create a new immutable version with the changed data, or to create a non-read-only model. And I have a problem with the idea of having to have separate readonly and non-readonly versions of what is essentially the same object for all sorts of reasons. 

I think I would prefer to be able to choose whether an instance of a specific object class is:
* Created fully frozen - by default or explicitly
* Frozen explicitly at a later point
* Optionally allowing full or partial unfreezing at a later point
* With the ability to do the same at an attribute level rather than a class level

PHP's readonly class or attribute designator is all or nothing. I think that to achieve the above you actually need object methods to validate attribute setting against the class or attribute frozen status (with readonly attributes obviously being permanently frozen).

**Comment by macropay-solutions** [comment]
@Sophist-UK Thanks for the insights. The idea started from  @cosmastech 's need of sending objects that must be readonly.
Using the Model for that is not fail-safe because you can freeze it and unfreeze it where you would not want to (in the part of the code that should only read it).
The conditional update feature was handled already a long time ago, see:
https://github.com/macropay-solutions/laravel-crud-wizard-free/blob/64c1dd1fe44a381a30a89b5be52d9aabba8d315d/src/Models/BaseModel.php#L31

**Reply by Sophist-UK** [reply]
I think that there needs to be the flexibility to allow or disallow freezing / unfreezing.

The point is that there are several different use cases, and it is probably easier to come up with a single solution based on a single object than start creating sub-classes of Models etc.

**Reply by macropay-solutions** [reply]
It is impossible to not overwrite any model implementation for this.
In other words it is possible to override any model implementation for this, thus making the solution breackable.

**Comment by ghost** [comment]
@macropay-solutions  The solution does not prevent the user from making another trip to db and fetch the model.

**Comment by macropay-solutions** [comment]
Merged the feature into the [free lib v3.4.9](https://github.com/macropay-solutions/laravel-crud-wizard-free/releases/tag/3.4.9) and also in both demos. Who will need it will be able to use it.
@bulletproof-coding we added that mention in readme.

**Comment by macropay-solutions** [comment]
Thanks [to this](https://medium.com/p/6bf274e341fa), this can be used for example in blades: 

```php
Route::get('/', fn (): View => \view('welcome', [
    /** Illuminate\Support\Collection<int, OperationFrozenAttributes> */
    'operations' => Operation::query()->limit(15)->get()
        ->map(fn (Operation $o): OperationFrozenAttributes => $o->getFrozen())
]));
```

**Reply by macropay-solutions** [reply]
Using [Lazy Object](https://www.php.net/manual/en/language.oop5.lazy-objects.php) in PHP ≥ 8.4
```php
    public function getFrozen(): OperationFrozenAttributes
    {
        $reflector = new ReflectionClass(OperationFrozenAttributes::class);
        $t = $this;

        return $reflector->newLazyGhost(function (OperationFrozenAttributes $object) use ($t): void {
            $object->__construct((clone $t)->forceFill($t->toArray()));
        });
    }
```

**Comment by macropay-solutions** [comment]
Thanks to [this unrelated discussion](https://github.com/laravel/framework/discussions/31778) we accidentally implemented another alternative for this frozen model that can be locked and unlocked for any updates @cosmastech :


```php
    /**
     * Prevent updates
     * Note that relations can be loaded and updated during the lock
     */
    public function lockUpdates(bool $checkDirty = true): bool
    {
        if (
            !$this->exists
            || $this->tmpDirtyIfAttributesAreSyncedFromCashedCasts !== null
            || ($checkDirty && $this->isDirty())
        ) {
            return false;
        }

        $this->tmpDirtyIfAttributesAreSyncedFromCashedCasts = [];

        return true;
    }

    /**
     * Unlock updates
     *
     * To reset the model's $attributes and get the changes from dirty applied during the lock use:
     *
     * if ($this->unlockUpdates()) {
     *  $dirty = $this->getDirty();
     *  $this->attributes = $this->original;
     *  $this->classCastCache = [];
     *  $this->attributeCastCache = [];
     * }
     *
     * Note that relations can be loaded during the lock
     */
    public function unlockUpdates(): bool
    {
        if ($this->hasUnlockedUpdates()) {
            return false;
        }

        $this->tmpDirtyIfAttributesAreSyncedFromCashedCasts = null;

        return true;
    }

    public function hasUnlockedUpdates(): bool
    {
        return $this->tmpDirtyIfAttributesAreSyncedFromCashedCasts !== [];
    }
```
see [commit](https://github.com/macropay-solutions/laravel-crud-wizard-free/pull/32/commits/ab31b0651be25cc69590bc927a511be6803bd8f4)

**Reply by marius-ciclistu** [reply]
https://marius-ciclistu.medium.com/prevent-updates-to-eloquent-model-vs-frozen-model-read-only-dto-72cea4c06aa3
